crates/mun_memory/src/gc/mark_sweep.rs

```diff
@@ -551,15 +798,20 @@ enum Color {
 
 /// An indirection table that stores the address to the actual memory, the type of the object and
 /// meta information.
-#[derive(Debug)]
 #[repr(C)]
 struct ObjectInfo {
-    pub ptr: *mut u8,
+    pub data: ObjectInfoData,
     pub roots: u32,
     pub color: Color,
     pub ty: Arc<TypeInfo>,
 }
 
+#[repr(C)]
+union ObjectInfoData {
+    pub ptr: *mut u8,
+    pub array: *mut ArrayHeader,
```

Comment:
Do all objects store the pointer now?

Reply:
No this is a `union`, so depending on the type stored in the ObjectInfo either the `ptr` or the `array` is valid.